### ISSUE TITLE
Require a finding on a contravariant parameter type? [working decision: no]

### ISSUE THREAD

**Issue by dzharkov** [jspecify/jspecify#49]
In the draft proposal, there's a part that allows for overrides to have value parameters with contravariant nullability:
> The method’s parameter types may have wider nullability than the corresponding superparameters' types, meaning, it suffices for superparameters to be type-convertible to the overriding method’s parameter types (while base types have to follow the normal rules of the compiler so overriding works as expected).

That might be a problem for the Kotlin compiler since in pure Kotlin value parameters are not contravariant in overrides and we apply the same logic when looking at the annotated Java code.

The reason we don't allow contravariant parameters in override is basically the same as in Java: they might be considered as overloads. In Java, one can not override with a wider-typed parameter (types should be equal). The same applies for Kotlin, but we have nullability as a built-in part of the type system.
```
interface A {
    fun foo(x: String)
}

class B : A {
    override fun foo(x: String) {
        
    }
    
    @JvmName("fooNullable")
    fun foo(x: String?) {} // is a different overload for `foo`
}

fun main() {
    B().foo("") // resolved to the not-nullable
    B().foo(null) // resolved to nullable
}

```

That means that when we see contravariant nullability in Java annotations, such code is assumed to be illegal thus we ignore annotations info for such parts completely as inconsistent:
```
interface A {
    void foo(@NotNull String x);
}

class B implements A {
    @Override
    public void foo(@Nullable String x) {

    }
}
```

`B::foo` has signature like `fun foo(x: String!)`, i.e. it has unknown nullability from Kotlin point-of-view.

But at the same time it's fine transforming from unknown nullness to both `NotNull` and `Nullable`:
```
interface A {
    void foo(String x, String y);
}

class B implements A {
    @Override
    public void foo(@Nullable String x, @NotNull String y) {

    }
}
```

In the above example `B::foo` is perceived by Kotlin as `fun foo(x: String?, y: String)`

**Comment by mernst** [comment]
Formal parameter type contravariance is consistent with Java, so it is natural and strongly desirable.  If we were designing Kspecify rather than Jspecify, we might make a different decision.

I think we should accommodate Kotlin whenever we can, but we should never degrade the behavior of Jspecify on Java code in order to conform to Kotlin's different semantics.

**Comment by kevinb9n** [comment]
I've kept coming back to this in my mind, and am just gonna reopen it for now. It may not be a critical priority right now, but I don't think that's what the open/close status is supposed to show.

If we don't have parameter contravariance, doesn't that mean that a method and all of its overrides can only have their annotations changed in perfect lockstep? That's against a project requirement, and it just feels bad to prohibit a method from accepting null if it wants to unless it changes the whole world around it.

**Comment by cpovirk** [comment]
(Reopening sounds fine.)

RE: changing annotations in lockstep: No, that's not required. See https://github.com/jspecify/jspecify/issues/49#issuecomment-514620654. Note also that, even with _specified_ nullness, library authors are likely to have the option to suppress the error on the overriding method. (But I grant that it's unfortunate to have to suppress in the case of a contravariant override (which is safe) just as you do for a covariant override (which is not).)

RE: consistency with Java: I can see how contravariance is:
- sound
- compatible with plain Java (since Java won't care about our annotations at all)

The one way in which it's arguably inconsistent is that Java doesn't support such contravariance at present. For example, I can't override `foo(String)` with `foo(Object)`: I just get an overload.

Now, again, overloading isn't possible in our case (since `foo(String)` and `foo(@Nullable String)` have the same erasure), so I'm not saying that users should be able to define two separate methods with those signatures. I just mean that, if users are able to override one with the other, we'd be somewhat breaking new ground.

I think we're all in agreement on this, but I wanted to make sure I'm not missing a deeper point that you're making.

Furthermore, breaking new ground might be perfectly fine. (I don't think I've formed a strong opinion yet. My main opinion so far is "not yet" :))

**Comment by kevinb9n** [comment]
re: "lockstep", right, I forgot that we were going to have to special-case unspecified. It would just be nicer to have the right thing happen for unspecified as a simple consequence of our general contravariant-parameters and covariant-returns support.

**Comment by kevinb9n** [comment]
BTW, in the case of covariant returns I don't think jspecify even needs to *do* anything; just by explaining how the annotations are meant to modify types, users "should" assume they can do covariant returns because that's just how return types work.

This one would need to be called out somehow though, because existing knowledge of types makes the wrong implication.

**Comment by cpovirk** [comment]
> It would just be nicer to have the right thing happen for unspecified as a simple consequence of our general contravariant-parameters and covariant-returns support.

Even with contravariant parameter types, we'd still like for unspecified to be considered to be "the same type" as both nullable and non-null.

The reason is parameterized types: Suppose a method in an unannotated interface accepts a type like `@Unspec List<@Unspec String>`. If we want to implement that in an annotated class, we would like to avoid `@Unspec`. Even if the right type is `@Nullable List<@Nullable String>`, that's not a supertype of the old method's parameters because parameterized types are invariant. So contravariance isn't enough.

But you have made me see that, just as the "same type" check can permit things that contravariant parameter types cannot, the reverse is also true: Contravariant parameter types can permit some declarations to work _without warnings or errors_, like if I override a method that accepts `@Unspec Object` with one that accepts `@Nullable Object`.

Still, I see "the same type" as the essential feature and contravariant parameter types as the possible add-on.

**Comment by kevinb9n** [comment]
The outcome of #111 may make this a moot point. But if we decide on the strict approach there we will still need to decide this. My argument that we should still support this either way (regardless of #111) is that there are already cases in the wild where a submethod accepts null even though its supermethod did not force it to. If we "don't support parameter contravariance" we will force that code to be annotated incorrectly.

**Comment by mernst** [comment]
Can we revisit this issue?

I propose that JSpecify should permit formal paremeter types to have wider nullability than the corresponding superparameters' types.

A specific tool might choose not to support formal parameter type contravariance, but I don't think the JSpecify *specification* should forbid it for all tools.

Arguments in favor:
 * This is behavioral subtyping, which is well-understood, sound, and flexible.
 * We have concrete examples of real-world code that benefits from this flexibility.
 * Tools such as IntelliJ IDEA and the Checker Framework support this.
 * There is no effect on overriding or overloading, which depend on Java basetypes.

Arguments against:
 * Maybe this complicates the Kotlin implementation.  (It's not clear to me from the discussion whether this is definitely true or might be true.)  I personally do not find this a compelling argument for  **J**Specify, even if it might be for **K**Specify.
 * Forbidding this permits us not to think about it.  I don't agree with this, because forbidding it is taking a stand, and to do so we need to think about and understand the implications.
 * The specification is slightly simpler if we forbid it.  This is true.

Are there other arguments for or against, or more compelling ways to phrase my summary?

In my view, the simplification is not worth the costs, which are reduced flexibility and expressiveness, and inconsistencies with implementations and the foundational underpinnings.

**Comment by cpovirk** [comment]
My view on this have continued to fluctuate.

Lately, my thinking has been that it's not necessarily our business to write a rule about this at all. I've come to like the idea that we lay out rules for subtyping\[*\], and we leave it up to tool authors to decide what to do with that. In general, I'd expect for tool authors to apply rules similar to those in the JLS: ["For example, the value of a return statement must be convertible to the method's return type."](https://github.com/jspecify/jspecify/blob/68a7dacd698c6b24e6cd57d37964f984a6512937/samples/README.md#syntax) But this is not any sort of obligation.

That gives tool authors leeway to apply in a case like this: The JLS would say "no contravariance." But the JLS is working within technical limitations that don't apply here, so there is an easy-to-understand case for tool authors to _not_ follow the JLS in this case.

(Previously in that spirit: [I make a case against writing a rule specifically about parameterized types.](https://github.com/jspecify/jspecify/issues/158))

Even if we decide not to officially rule one way or the other on this, I appreciate everyone's contributions to to understanding the tradeoffs of contravariance. That can help tool authors make their own decisions. To build on a couple things Mike said:

On a downside of contravariance: It's possible that users will find that Kotlin ignores their annotations. (Of course, this can happen, anyway, if they suppress errors or don't run a nullness checker at all!) Or, if the Kotlin team does feel this is worth changing, then their users may see changes in the behavior of their existing code (unless the Kotlin team customizes the behavior only for JSpecify annotations and not for other nullness annotations). In theory, we could even collect data about how often these situations arise.

On an upside of contravariance: Another kind of data that could be helpful is cases in which contravariant parameter types have come up in practice, taken from users of Checker Framework, IntelliJ, etc. (Some of this may have come up in previous docs that I don't have handy. The one example I recall mentioning up-thread is `Collection.contains`, though that depends on resolving another unsettled discussion.) I'm hoping that early testers can give us additional data on this.

As always, if anyone else knows of tradeoffs, please do add them to the list.

\[*\] and substitution, and whether two types are "the same," and "defaulting," and what an unbounded wildcard means, and... -- but basically to restrict ourselves to "how to know what type something has" and "how to perform a test on two types"

**Comment by kevinb9n** [comment]
The Kotlin discussion seemed to be just about what Kotlin currently does with existing annotations, but I don't see a clear implication that there would be major hurdles to adopting our semantics. I assume those challenges are surmountable (especially given that IntelliJ's inspections already allow this scenario), but corrections would be welcome.

(Remember, if we can get agreement on https://github.com/jspecify/jspecify/issues/111#issuecomment-862976659 then that will autoresolve this one too.)

**Comment by kevin1e100** [comment]
The root concern for Kotlin as I understand it is that in Kotlin, a class can declare _both_ `fun foo(x: String)` and `fun foo(x: String?)`, as they're considered distinct overloads. Because of that, if a class A declares `fun foo(x: String)`, then a class B that extends A *cannot override* A's `foo` as `fun foo(x: String?)`, because that would be a different method.  Instead, the override must be declared `fun foo(x: String)`, ie., identical to A (modulo return type, which I believe can covariantly change).

Now imagine A was declared in Java with `@NullMarked` and B in Kotlin. It's hard to see how B could override A's `public void foo(String x)` with `fun foo(x: String?)` without confusion, especially if it _also_ wants to declare `fun foo(x: String)`. Now I think it could be done/allowed (since one of the `foo` functions is required to use `@JvmName` so it has a unique name in bytecode), but it seems difficult to do correctly unless the same is somehow allowed if both A and B are declared in Kotlin.

Let's quickly think about the inverse, i.e., A declared in Kotlin and B in Java.  That seems fine. However, a class C declared in Kotlin that extends B now has to be careful. Without thinking through too much it it seems it may be safe to ignore A's declaration and just inherit B's (or ignore B's and inherit A's actually, since it's stricter), but it would certainly be wrong to inherit both.

**Comment by mernst** [comment]
I propose that we resolve this issue (and also others such as https://github.com/jspecify/jspecify/pull/169) in the following way:
JSpecify permits standard behavioral subtyping.

Unless I have missed some subtleties in the long discussions, everyone agrees that that behavior is desirable for specifying Java code.
It's also what is already implemented by widely-used tools, and what users of those tools want.

There may be an incompatibility with current Kotlin behavior, but this project is **J**Specify, not **K**Specify.

**Comment by cpovirk** [comment]
I would want to think more about this before declaring that it's _definitely_ desirable for Java code to support this. My initial reaction is that it _would_ definitely be desirable if all users ran a nullness checker. But in today's world, I could imagine problems: If users happen to use `MyFoo.doSomething(@Nullable String)` today, they could pass `null` safely. But if they switched to `SomeOtherFoo.doSomething(@NonNull String)`, their code would start to fail. For example, Guava's [`CharStreams.asWriter`](https://guava.dev/releases/snapshot-jre/api/docs/com/google/common/io/CharStreams.html#asWriter(java.lang.Appendable)) has been returning a `Writer` whose `write(String)` method accepts `null`. This is [not how most other `Writer` implementations behave](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/io/Writer.html#write(java.lang.String)), so [we changed it](https://github.com/google/guava/commit/7112c0339419289cf7175f304d092571c5913ec2).

On the other hand, I would assume that contravariant parameter types would provide value to some JSpecify users, too, just as covariant return types do. I would just feel most comfortable if we had some more examples before I committed too strongly to either position.

As for Kotlin, I think we've heard from a few different organizations that they are interested in Kotlin's interaction with JSpecify. Interoperability between JVM languages is a nice feature of Java, but certainly most users of Java are Java, so we'll want to weight concerns accordingly.

Fortunately, in this case, I _think_ Kotlin users aren't actually practically hurt by Kotlin's current behavior: Whether Kotlin sees a method as accepting `Foo?` or `Foo!` is almost entirely irrelevant to users: Kotlin will allow users to pass `null` in either cases. Someone please correct me if I'm wrong, but if I have that right, I don't see much harm from Kotlin's current behavior if people declare methods with contravariant parameter types.

Put that all together, and at a minimum, I continue not to see value in saying that parameter contravariant is illegal/banned/disallowed. Now, I also still want to think more about the line between what we specify and what we don't. As a result, I am nervous about including _normative_ comments about behavioral subtyping: Once we start saying that _x_ and _y_ are permitted, does that mean that _z_ is _not_ permitted? And what _exactly_ is covered by _x_ and _y_? I am going to try to open a separate issue about this more general concern in the coming days, though, since this current issue touches on some concerns specific to parameter types.

(As a short-term patch to #169, I am still happy to remove the sample. But I understand that you ultimately want to see the underlying issue resolved.)

**Comment by cpovirk** [comment]
I have filed #206 about what I see as one of the underlying principles. I'd be interested in whether you think that it's important to make "JSpecify permits standard behavioral subtyping" part of the normative specification. In particular, I'm pondering whether it's sufficient (assuming that it's even desirable!) to call questions like "valid parameters for overrides" explicitly _not formally specified_, while still explicitly _non-normatively_ encouraging authors of nullness checker to support behavior subtyping when desired.

**Comment by cpovirk** [comment]
Setting aside my concerns [around the underlying principle of how much we normatively specify](https://github.com/jspecify/jspecify/issues/49#issuecomment-878453162), I should present a new(?) example that may show parameter contravariance at its best:

```
interface Multimap<K extends @Nullable Object, V extends @Nullable Object> {
  Collection<V> get(K key);
  ...
}

class ImmutableMultimap<K, V> implements Multimap<K, V> {
  ImmutableCollection<V> get(???);
}
```

Background: The reason that `Multimap.get` _must_ require a `K`, rather than [accepting an `Object` as methods like `Collections.contains` do](https://errorprone.info/bugpattern/CollectionIncompatibleType), is that `get` returns a live view that supports `add(V)`. Thus, if someone calls `get(notK).add(v)`, our multimaps may insert a wrongly typed key. (_Sometimes_ multimaps would at least have the ability to detect and reject the wrongly typed key at runtime, but we can't completely guarantee that. And anyway, runtime failures are unfortunate.)

(Incidentally, if you ever do want to retrieve the values associated with a key that might not statically be a `K`, our general recommendation is to call `multimap.asMap().get(key)`, noting that the call will return `null` instead of an empty collection if the key is not present.)

`ImmutableMultimap.get` is different: Because the multimap is immutable, there is no danger of `get(notK).add(v)`. Thus, `get` could sensibly accept `@Nullable Object`. Of course, Java won't let us change from `K` to `Object`. But parameter contravariance for nullness would at least let us declare the method as accepting `@Nullable K`!

How much does that matter? Conveniently, I recently replaced `@Nullable K` with `K` on (all?) our `Multimap.get` methods. (Previously, we'd been applying `@Nullable` to type-variable usages according to a non-JSpecify rule.) That change did break some users' nullness tests. (Well, we didn't _break_ them: We added suppressions :))

In most cases, the users were working with mutable multimaps, so the change was closing a real type hole. (In reality, most users probably never wrote to the map, so I doubt we were detecting many "real bugs," if any.)

That left maybe a third (approximately, and the sample was small to begin with) who were working with immutable multimaps. Those users were definitely in no danger of type holes. The only possible danger is passing `null` when the code "should" have dealt with `null` in some other way, and that's similar to [the general danger with methods like `Collection.contains`](https://errorprone.info/bugpattern/CollectionIncompatibleType), which we live with.

My impression from looking very briefly at those callers is that none passed `null` in practice. Often, they were looking up a key that they'd definitively inserted into the multimap. Still, `get(K)` forces them to do _something_ (check `!= null`, call `castNonNull`, suppress, ...), while `get(@Nullable K)` would not.

**Comment by kevinb9n** [comment]
Friendly reminder that if we decide we can accept my proposal in #111, then this issue becomes technically a non-issue for JSpecify, though we can still have advice for tool owners on a voluntary basis.

Just a quick incomplete thought:

Let's look at Java's standard behavior a sec: With `Foo x(Bar bar) {}`, I'm allowed to override it with `SubFoo x(Bar bar) {}`, but I can't override it with `Foo x(SuperBar bar) {}`.

So here's a hot take: actually I kind of *can*, it's just that I have to write the glue myself: `Foo x(Bar bar) { return x((SuperBar) bar); }`.

But in JSpecify-land, you can't do that. So this becomes one possible valid reason for diverging from the Java precedent.

**Comment by kevin1e100** [comment]
Apologies, but could you link directly to the comment in #111 you're referring to that makes this here issue "technically a non-issue"? I'm honestly very confused, since #111 seems to talk about handling parameter _co_variance, while this issue appears to be about parameter _contra_variance.

I don't mean to sound like a broken record, but the concern with allowing parameter contravariance, as alluded to in the [issue description](https://github.com/jspecify/jspecify/issues/49#issue-471586073), is to my understanding purely about compatibility with Kotlin; the concern is _not_ whether or not parameter contravariance is safe (yes) or useful (yes, though how commonly is unclear to me). 

@petukhovv do you think Kotlin could be made to deal with contravariance in Java and mixed Java/Kotlin method parameter nullness? Seemingly this could be done since:
* Kotlin method over_loads_ with differing nullnesses of otherwise identical method parameters have to have differing JVM names, IIUC, so overriding any overload in a hypothetical Java subclass should be unambiguous (if confusing).
* By the same notion, overriding a Java-defined method in a Kotlin subclass with contravariant parameter nullness seems likewise unambiguous.

It does seem like it could be significant work to correctly design and implement such a mechanism for Kotlin, even if it's technically doable. Do you think it would be feasible for you to tackle this?

**Comment by petukhovv** [comment]
@kevin1e100 sorry for the delayed answer. I think we could treat some additional cases related to mixed nullness, but I still don't fully understand what kind of reaction from Kotlin is expected for what specific cases. Is it some kind of warnings? Could you please give a couple of examples with Java/Kotlin code snippets and the wanting reaction from the Kotlin compiler on them?

**Comment by kevinb9n** [comment]
> Apologies, but could you link directly to the comment in #111 you're referring to that makes this here issue "technically a non-issue"? I'm honestly very confused, since #111 seems to talk about handling parameter _co_variance, while this issue appears to be about parameter _contra_variance.

Ah yeah, I see how confusing that is, now. Please ignore this for now until I restate #111 better.

**Comment by kevinb9n** [comment]
> I don't mean to sound like a broken record, but the concern with allowing parameter contravariance, as alluded to in the [issue description](https://github.com/jspecify/jspecify/issues/49#issue-471586073), is to my understanding purely about compatibility with Kotlin;

Sorry I've seemed to be ignoring that. I admit I don't fully understand it, but have been expecting that someone who isn't me would be providing the "here's how we know it's a total deal-breaker" comment at some point, and we can still figure out what we would _otherwise_ do short of that.

EDIT: Chris says: "I think that the Future you get from ScheduledExecutorService.scheduleAtFixedRate could be a Future<@NonNull Void>: It is a repeating task, so it never completes successfully. get() either hangs or throws." So... while we still don't have to special-case Void at all, if we do it but don't have `@NonNull` then we'd actually be taking away something that a tiny number of APIs *might* have actually wanted to do. Meh!

**Comment by kevinb9n** [comment]
That was a borked comment because I pasted to the wrong place - ignore it.

**Comment by kevinb9n** [comment]
Check my understanding?

In null-marked context, `class SuperFoo { void bar(String x) {} }` (or, I believe, even the equivalent class written in Kotlin). I think:

* When a _Java_ class extends SuperFoo, Java's subclassing rules are in effect, so a `bar(String)` method overrides the one above, however its parameters might be annotated, and only the subclass annotations are relevant to consumers of its specific static type.
* When a _Kotlin_ class extends SuperFoo, Kotlin's rules are in effect. It has 2 independent choices: override-with-same-nullness or not; overload-with-other-nullness (using a JvmName) or not. It doesn;'t offer override-and-change-nullness in either direction.

**Comment by kevinb9n** [comment]
**Proposal (reaction icons appreciated):**

The case looks strong for a solid working-decision to me:

* I just can't see any motivation _JSpecify_ would have for objecting to such annotations.
* In fact, there is *some* existing code where this would be the only correct way to annotate it. (Likely not enough to really demand action on its own, but it doesn't really _need_ to.)
* The Kotlin behavior described at top seems fixable and like it _should_ be fixed (pay no mind to annotations on the *overridden* method), BUT,
* Even if not, it would only justify a response of "we'll document why you might not want to do this", not "so we'll ban it for everyone". A problem with a single tool should have to be _very_ serious in order to force the decision on everyone.
* In the other direction, when Kotlin embeds our annotations in its class files, it simply won't need/use this feature, so I don't see a problem here either.

I stress that this decision would do nothing to cut off conversation about how Kotlin can best deal with this, nor would it prevent a new issue from assembling a proper claim that this would be so devastating to Kotlin as to justify a ban.

Thoughts (or just thumbs)?

**Comment by kevinb9n** [comment]
> * I think that, practically speaking, there's minimal harm in letting this pass without a warning/error (in particular, minimal harm to Kotlin users), and I'd want tools to be free to do so.

They'd have that freedom either way. But a "no" on my proposal would mean that (to be compliant) they still have to prove they can recognize this case as such. There could be some value in that, but I think it's not worth worrying about for the foreseeable future.


> * I don't think the feature is 100% benefit; I see [a small _possible_ cost to users who do _not_ use nullness checkers](https://github.com/jspecify/jspecify/issues/49#issuecomment-878342831).

Fair enough, but seems more like "why you might not want to do this" advice.


> * I am nervous about getting into rules about specific language constructs unless we have to ([#206](https://github.com/jspecify/jspecify/issues/206)): In general, I like the idea that Java has a type system, to which JSpecify-the-spec is just adding nullness, not adding further enhancements. I know that's another discussion that we'll be picking up.

That is a "base" model I still strongly agree with. But this isn't about "type errors", it's just about override resolution.
There is a precedent: Java *could* already allow `m(Iterator<? extends SuperFoo>)` to override `m(Iterator<? extends SubFoo>)` (couldn't it?) but it doesn't.

This feels inconsistent with that, but my argument would be: override resolution is definitively none of our business! It's already done and unchangeable by the time we wake up. And of course, we have special considerations because of legacy code.


> * If some tool then chooses to produce a warning/error for parameter contravariance, I don't think it's a huge problem for the subclass to have to suppress a warning on the probably rare occasion that this comes up.

I agree with all this and I don't think it conflicts with my proposal.  I am comfortable with simply "hoping" that such a tool would word that warning veeerrry gently, for example saying "but if this accurately models what the code is doing, then this is a false alarm".

So my proposal remains. **Please react**

**Comment by kevinb9n** [comment]
Thanks for reopening, I absolutely read your points as registering mild asterisks and not as major concern.

**Comment by cpovirk** [comment]
Kevin and I talked some offline yesterday. We didn't get to the bottom of everything, but here are some thoughts:

I don't think that anyone is currently pushing for the idea that:
- "If a nullness checker tells users about 'nullness mismatches' but does _not_ tell them about contravariant parameters, then that checker is somehow 'lacking' from the perspective of JSpecify."

(It _might_ also be that no one is currently pushing for the idea that a checker that _does_ tell users about contravariant parameters is "lacking" from the perspective of JSpecify. But even if that's the case, I do think we'd advise checker authors to make clear that contravariant parameters do not introduce a _soundness problem_.)

I think that we could title this issue in some way to reflect that and then close it ("If we take a stance on contravariant parameters, will it be against them \[no\]?"). (If we additionally want to come to a conclusion about what to say about a tool that _does_ tell users about contravariant parameters, then we should float a proposed policy on that, whether on this issue or a new one.)

I think that most of the controversy on this thread has been around related claims, like:
- Without parameter contravariance, unspecified nullness won't work well with overloads. ([I don't think that's the case.](https://github.com/jspecify/jspecify/issues/49#issuecomment-643383260))
- It would be questionable for a tool to apply JSpecify rules to parameters of overrides. ([I think it would be desirable for it to do so.](https://github.com/jspecify/jspecify/issues/49#issuecomment-1043167348))
   - I think this is part of what's behind...
     > But a "no" on my proposal would mean that (to be compliant) they still have to prove they can recognize this case as such.

     I think Kevin is saying that we have a choice between "no samples / conformance tests around parameters" and "some samples / conformance tests around parameters," and he would prefer none because that puts less obligation on tool authors. But I think that we would want to cover cases of parameter checking no matter what (e.g., _covariant_ parameters). Plus, tool authors would want to make a decision on whether to tell users about contravariance or not, so they'd want to test that case, anyway.
- \<other general questions around defining "JSpecify compliance," touching on [#232](https://github.com/jspecify/jspecify/issues/232) (which I still need to reply to :( ...) and perhaps [#206](https://github.com/jspecify/jspecify/issues/206)\>

If one of those claims is more closely tied to this issue than I think (e.g., if people are in favor of contravariant parameters only because they think it's necessary for unspecified nullness to work with overrides), then we should get to the bottom of that. But I'm hoping that there's no actual controversy on the "Would we come out against contravariant parameters?" question. In that case, I think we can address the more controversial questions separately and more directly.

**Comment by kevinb9n** [comment]
For future reference, I this issue is about what happens at "// A": we expect a finding, we expect no finding, or we don't care.

```
static class SuperFoo {
  void bar(String s) {}
}
static class SubFoo extends SuperFoo {
  @Override void bar(@Nullable String s) {} // A
}
void test(SuperFoo sup, SubFoo sub) {
  sup.bar(null); // B
  sub.bar(null); // C
}
```

**Comment by cpovirk** [comment]
(Kevin and I are going to talk synchronously about this one of these days, I promise!)

I think the disagreement is indeed precisely over what I see as a pivot in that framing: What I think most people most want to know is whether nullness checkers are allowed to let `// A` pass without some kind of warning or error to the user:
- That lets users know that they can probably write code like that without having to fight the compiler.
- That lets tool authors know that they can provide such tools to users who want them.
- That lets tool authors know that _interoperating_ with code that was annotated that way is something that they'll likely want to think about, even in a far-off, hypothetical, fully annotated world.

I think we all agree (but anyone who disagrees, please speak up!) that the answer is "Yes, tools can let that pass \[in no small part because they can let _anything_ pass!\]."

I think the second most interesting question is whether we'd _discourage_ that because of some kind of concern (like [the Kotlin platform-type concern](https://github.com/jspecify/jspecify/issues/49#issue-471586073) or [the non-nullness-checked-user concern](https://github.com/jspecify/jspecify/issues/49#issuecomment-878342831)). I think that there's at least general consensus behind the idea that we wouldn't even _discourage_ it.

The "finding" framing is a third question for us to talk about. But it seems to be purely a question about what kind of test tool authors would find more convenient. I think that tool authors would be happy with the answer "We should provide tests that declare contravariant parameters and _don't_ identify them as mismatches." But I think we should be very clear in distinguishing these three questions, and I think we should make sure we're answering the first two and not just the third.

**Comment by kevinb9n** [comment]
To your three questions:

1. Yes, we agree
2. Yes, we agree
3. Should the conformance test suite include a contravariant-parameter scenario? My first preference would be "yes, and it should test that NO override-conflict finding is produced." A tool can still give its own warning for its own reasons, as always, but that's its own deal and not a "JSpecify thing". Does that work? If necessary would settle for my second preference, "no".... just want to get this resolved.

**Comment by netdpb** [comment]
**Current decision**: If an overriding method has a nullable parameter where the method it overrides has a non-null parameter, JSpecify will not require that tools report a problem to users. It may require that tools demostrate that they can recognize that case.

**Argument for changing**: Kotlin does not allow such overrides. However, JSpecify focuses on Java declarations.

**Timing**: This must be decided before version 1.0 of the spec.

**Proposal for 1.0: Finalize the current decision.** If you agree, please add a thumbs-up emoji (👍) to this comment. If you disagree, please add a thumbs-down emoji (👎) to this comment and briefly explain your disagreement. Please only add a thumbs-down if you feel you can make a strong case why this decision will be materially worse for users or tool providers than an alternative.

**Comment by kevinb9n** [comment]
Unfortunately I don't think the statement there is right and we should suspend voting a sec. I'm also reopening until we agree on exactly what this issue is about.

This issue is about something devilishly subtle. The root of the confusion is that, while a non-null parameter overriding a nullable one is *an issue*, that still might be the absolutely correct way to annotate the API, so if it encouraged the user to change the annotations it would be doing some actual harm.

The decision I was seeking here was that we will actually test that tools *do not* produce a finding in this case. If a tool does want to produce a finding, it's doing so for its own reasons, not for "JSpecify reasons" (hard to define; sorry). So, just like the tool's "bad variable name" finding or whatever, it shouldn't be enabled when running our conformance tests.

An alternative option would be that it is its own category of JSpecify finding different from (all?) the others. I think that would probably be okay too, just maybe unnecessary?

**Comment by cpovirk** [comment]
I keep feeling that I'm almost on top of this issue, only for things to get slippery again.

JSpecify currently says essentially nothing about when its subtyping check and similar checks should be deployed on Java source. This issue might be about starting to say such things, or it might be about what conformance tests we write.

Part of my concern has been that we both seem to want to say "We don't tell checkers when to apply or not apply our checks" and also seem to want to say "But by writing a conformance test about this situation, we are implicitly telling checkers that applying our checks here would be in some way 'unjustified.'"

The samples tried to [dodge this question](https://github.com/jspecify/jspecify/blob/main/samples/README.md#syntax) by saying that they're written under the assumption of JLS-like rules (which of course ban contravariant parameters) but that that those rules (a) have no necessary connection to what checkers do and (b) are more strict than "necessary":

> You can think of these as extending the normal JLS type rules to cover types that have been augmented with nullness information. (For example, the value of a return statement must be convertible to the method's return type, and the receiver in a method call should not be `@Nullable`.) ... Nullness checkers are... not _obligated_ to do anything for any particular comment.... Note that a nullness checker can be sound even if it does not issue errors for some cases of `jspecify_nullness_mismatch` or `jspecify_nullness_not_enough_information`!

I would like more clarity on whether this issue is "We will not write a conformance test for this case" or "We will write a conformance test for this case that contains no cannot-convert comment" or "We will recommend that tools not produce errors in such situations" or some combination of those or what.

I also wonder a little if we're opening a can of worms by trying to go beyond the JLS rules for when to apply checks. For example, if a supermethod has a type parameter `<T extends Object>`, is it OK for a submethod to declare it as `<T extends @Nullable Object>` _if_ the resulting changes produce only covariant return-type changes and contravariant parameter-type changes? Maybe we don't need to care about that relative edge case. I just think I recall some complexity in the Checker Framework that seemed to arise from trying to handle such overrides. So I'd also like to have some idea of whether we want to say something like "We've thought specifically about contravariant parameter types, and we think they work well in practice" or "We think that going beyond the JLS is safe in general."

I may be making a big thing out of an inconsequential decision. Part of my concern is that it keeps _sounding_ inconsequential, only for phrasing about what is (e.g.) "correct" to crop back up.

If I had managed to collect my thoughts earlier, I would have talked to kevinb9n about this in person.... We should probably talk sometime.

**Comment by cpovirk** [comment]
tl;dr I think that "finding" and "JSpecify-based" are undefined terms that are making the question slippery.

My proposal would be:
- settle now on one of the options for [what kind of conformance test to write](https://github.com/jspecify/jspecify/issues/49#issuecomment-1162379463)
- postpone [further discussion of deeper issues that have remained unsettled](https://docs.google.com/document/d/1k82nDt28-sKidiDsN1vSOvZQuJkH12jbHiB5jO7o2z8/edit?resourcekey=0-kH3OFiGZnPHYz3tIgBuw1Q)

If it helps, I would also be open to also agreeing to expand some non-normative resources to explain why contravariant parameters are worth considering.

**Comment by cpovirk** [comment]
After some discussion elsewhere, I think I could get behind [the proposal](https://github.com/jspecify/jspecify/issues/49#issuecomment-2045964905) if we change "'JSpecify-based' finding" to "JSpecify conformance finding," which I think makes clearer that this is about what the checker must be able to recognize internally in order to pass our tests. (There are surely other formulations that could work, too, but that one might minimize the edit distance :))

**Comment by cpovirk** [comment]
Fun fact that I can't resist posting here: In one narrow way, you _totally can_ declare a contravariant parameter type in Java:

```java
class OverrideNoTypeVariable {
  interface Super<T> {
    void x(T t);
  }

  interface Sub<T> extends Super<T> {
    @Override
    void x(Object t);
  }
}
```

It builds without even a warning.

(["the signature of m1 is the same as the erasure (§4.6) of the signature of m2."](https://docs.oracle.com/javase/specs/jls/se22/html/jls-8.html#jls-8.4.2))

**Comment by cpovirk** [comment]
I just tried the contravariant parameter type with `ImmutableMultimap.get`, since it could be nice for it to accept `Object` instead of `K` (though, if we do that, we should probably update https://errorprone.info/bugpattern/CollectionIncompatibleType accordingly :)).

That turns out to not go so great: Calls that pass a `K` appear to start resolving to the supermethod `Multimap.get`, whose return type is `Collection` instead of `ImmutableCollection`, leading to build errors in code that expects `ImmutableCollection`.

**Comment by netdpb** [comment]
**Proposal**: As described in [JSpecify Conformance: Questions and Answers](https://github.com/jspecify/jspecify/wiki/JSpecify-Conformance:-Questions-and-Answers), conformance tests will check that tools can recognize when a parameter's type is nullness-contravariant with respect to its supermethod, but that will impose no requirement on tools to report anything in particular to users.

**Timing:** This must be decided before version 1.0 of the spec.

If you agree, please add a thumbs-up emoji (👍) to this comment. If you disagree, please add a thumbs-down emoji (👎) to this comment and briefly explain your disagreement. Please only add a thumbs-down if you feel you can make a strong case why this decision will be materially worse for users or tool providers than an alternative.

**Comment by kevinb9n** [comment]
This seems different from what Chris and I were approaching agreement on, but I am considering just agreeing to get this over with, and it probably wouldn't be a disaster.

**Comment by cpovirk** [comment]
Is the difference just between...

- "This is a different case than non-contravariant parameter changes, so our conformance tests will require tools to distinguish those cases by expecting no finding in this case."
- "This is a different case than non-contravariant parameter changes, so our conformance tests will require tools to distinguish those cases by expecting a different kind of finding in this case."

...or is there anything deeper going on? Either way, we'll probably want to take it to a doc, as you said.
